<a href="https://colab.research.google.com/github/bbchen33/Data-Science/blob/master/Genome_data_wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Data wrangling for genome and genes from an unfavored organism - *Clostridium thermocellum*
First, the annotated genome is downloaded from NCBI.

In [2]:
!wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/184/925/GCF_000184925.1_ASM18492v1/GCF_000184925.1_ASM18492v1_genomic.gbff.gz

--2019-10-19 21:45:59--  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/184/925/GCF_000184925.1_ASM18492v1/GCF_000184925.1_ASM18492v1_genomic.gbff.gz
           => ‘GCF_000184925.1_ASM18492v1_genomic.gbff.gz’
Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.11, 2607:f220:41e:250::13
Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.11|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /genomes/all/GCF/000/184/925/GCF_000184925.1_ASM18492v1 ... done.
==> SIZE GCF_000184925.1_ASM18492v1_genomic.gbff.gz ... 2437250
==> PASV ... done.    ==> RETR GCF_000184925.1_ASM18492v1_genomic.gbff.gz ... done.
Length: 2437250 (2.3M) (unauthoritative)

GCF_000184925.1_ASM 100%[===================>]   2.32M  12.7MB/s    in 0.2s    

2019-10-19 21:46:00 (12.7 MB/s) - ‘GCF_000184925.1_ASM18492v1_genomic.gbff.gz’ saved [2437250]



In [3]:
!ls

GCF_000184925.1_ASM18492v1_genomic.gbff.gz  sample_data


In [0]:
!gunzip GCF_000184925.1_ASM18492v1_genomic.gbff.gz

In [5]:
!head -100 GCF_000184925.1_ASM18492v1_genomic.gbff

LOCUS       NC_017304            3561619 bp    DNA     circular CON 03-SEP-2017
DEFINITION  Hungateiclostridium thermocellum DSM 1313, complete sequence.
ACCESSION   NC_017304
VERSION     NC_017304.1
DBLINK      BioProject: PRJNA224116
            BioSample: SAMN00713572
            Assembly: GCF_000184925.1
KEYWORDS    GSC:MIGS:2.1; RefSeq.
SOURCE      Hungateiclostridium thermocellum DSM 1313
  ORGANISM  Hungateiclostridium thermocellum DSM 1313
            Bacteria; Firmicutes; Clostridia; Clostridiales;
            Hungateiclostridiaceae; Hungateiclostridium.
REFERENCE   1  (bases 1 to 3561619)
  AUTHORS   Feinberg,L., Foden,J., Barrett,T., Davenport,K.W., Bruce,D.,
            Detter,C., Tapia,R., Han,C., Lapidus,A., Lucas,S., Cheng,J.F.,
            Pitluck,S., Woyke,T., Ivanova,N., Mikhailova,N., Land,M.,
            Hauser,L., Argyros,D.A., Goodwin,L., Hogsett,D. and Caiazza,N.
  TITLE     Complete genome sequence of the cellulolytic thermophile
            Clostridium thermoce

In [6]:
import pandas as pd
table = pd.read_table('GCF_000184925.1_ASM18492v1_genomic.gbff', header = None)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\t'.
  


In [7]:
table.head()

0
0  LOCUS       NC_017304            3561619 bp   ...
1  DEFINITION  Hungateiclostridium thermocellum D...
2                              ACCESSION   NC_017304
3                            VERSION     NC_017304.1
4                DBLINK      BioProject: PRJNA224116

In [8]:
table[table[0].str.contains('Annotation Date')].index

Int64Index([96], dtype='int64')

In [9]:
table.head(150)

0
0    LOCUS       NC_017304            3561619 bp   ...
1    DEFINITION  Hungateiclostridium thermocellum D...
2                                ACCESSION   NC_017304
3                              VERSION     NC_017304.1
4                  DBLINK      BioProject: PRJNA224116
5                              BioSample: SAMN00713572
6                            Assembly: GCF_000184925.1
7                    KEYWORDS    GSC:MIGS:2.1; RefSeq.
8    SOURCE      Hungateiclostridium thermocellum D...
9      ORGANISM  Hungateiclostridium thermocellum D...
10               Bacteria; Firmicutes; Clostridia; ...
11               Hungateiclostridiaceae; Hungateicl...
12                 REFERENCE   1  (bases 1 to 3561619)
13     AUTHORS   Feinberg,L., Foden,J., Barrett,T.,...
14               Detter,C., Tapia,R., Han,C., Lapid...
15               Pitluck,S., Woyke,T., Ivanova,N., ...
16               Hauser,L., Argyros,D.A., Goodwin,L...
17     TITLE     Complete genome sequence of the ce...
18                    Clostridium thermocellum DSM1313
19     JOURNAL   J. Bacteriol. 193 (11), 2906-2907 ...
20                                   PUBMED   21460082
21                 REFERENCE   2  (bases 1 to 3561619)
22     AUTHORS   Lucas,S., Copeland,A., Lapidus,A.,...
23               Pitluck,S., Davenport,K., Detter,J...
24               Hauser,L., Kyrpides,N., Ivanova,N....
25                  Foden,J., Feinberg,L. and Woyke,T.
26             CONSRTM   US DOE Joint Genome Institute
27     TITLE     Complete sequence of Clostridium t...
28                               JOURNAL   Unpublished
29                 REFERENCE   3  (bases 1 to 3561619)
..                                                 ...
120                     ##Genome-Annotation-Data-END##
121                         COMPLETENESS: full length.
122           FEATURES             Location/Qualifiers
123                         source          1..3561619
124                       /organism="Hungateiclostr...
125                            /mol_type="genomic DNA"
126                                 /strain="DSM 1313"
127                       /culture_collection="DSM:...
128                            /db_xref="taxon:637887"
129                            /collection_date="1926"
130                       /collected_by="Viljoen et...
131                          gene            212..1543
132                       /locus_tag="CLO1313_RS00010"
133                       /old_locus_tag="Clo1313_0...
134                          CDS             212..1543
135                       /locus_tag="CLO1313_RS00010"
136                       /old_locus_tag="Clo1313_0...
137                       /inference="COORDINATES: ...
138                       sequence:RefSeq:WP_020457...
139                       /note="Derived by automat...
140                       gene prediction method: P...
141                                     /codon_start=1
142                                   /transl_table=11
143                       /product="chromosomal rep...
144                       /protein_id="WP_003516465.1"
145                       /translation="MNTQLNEIWQK...
146                       AVPAEFNKGILESRYQTLIKNAIKQ...
147                       PLSVLNPKYTFDTFVIGNSNRFAHA...
148                       GHYILEQNSSQRVLYVSSEKFTNEL...
149                       ERTEEEFFHTFNALYEANKQIILSS...

[150 rows x 1 columns]

It looks like from row 0 down to row 130 is just the background information. Let's remove them.

In [0]:
table = table.iloc[131:]

In [11]:
table.head(30)

0
131                          gene            212..1543
132                       /locus_tag="CLO1313_RS00010"
133                       /old_locus_tag="Clo1313_0...
134                          CDS             212..1543
135                       /locus_tag="CLO1313_RS00010"
136                       /old_locus_tag="Clo1313_0...
137                       /inference="COORDINATES: ...
138                       sequence:RefSeq:WP_020457...
139                       /note="Derived by automat...
140                       gene prediction method: P...
141                                     /codon_start=1
142                                   /transl_table=11
143                       /product="chromosomal rep...
144                       /protein_id="WP_003516465.1"
145                       /translation="MNTQLNEIWQK...
146                       AVPAEFNKGILESRYQTLIKNAIKQ...
147                       PLSVLNPKYTFDTFVIGNSNRFAHA...
148                       GHYILEQNSSQRVLYVSSEKFTNEL...
149                       ERTEEEFFHTFNALYEANKQIILSS...
150                       IAILRKKAQLENLTVPNEVIVFIAD...
151                       EALKDILSANKAKVLNCTTIQEAVA...
152                       TEMSLPKIGEEFGGRDHTTVIHACE...
153                         gene            1793..2893
154                       /locus_tag="CLO1313_RS00015"
155                       /old_locus_tag="Clo1313_0...
156                         CDS             1793..2893
157                       /locus_tag="CLO1313_RS00015"
158                       /old_locus_tag="Clo1313_0...
159                       /inference="COORDINATES: ...
160                       sequence:RefSeq:WP_003513...

I only want to keep the locus tags and the actual amino acid sequence so I'll delete the rest of the rows.

In [0]:
rows_to_remove_list = ['gene','CDS','/inference','sequence:','/note','gene prediction','/codon_start','transl_table','/product=','protein_id']

In [0]:
for item in rows_to_remove_list:
  table = table[~table[0].str.contains(item)]

In [14]:
table.head(50)

0
132                       /locus_tag="CLO1313_RS00010"
133                       /old_locus_tag="Clo1313_0...
135                       /locus_tag="CLO1313_RS00010"
136                       /old_locus_tag="Clo1313_0...
145                       /translation="MNTQLNEIWQK...
146                       AVPAEFNKGILESRYQTLIKNAIKQ...
147                       PLSVLNPKYTFDTFVIGNSNRFAHA...
148                       GHYILEQNSSQRVLYVSSEKFTNEL...
149                       ERTEEEFFHTFNALYEANKQIILSS...
150                       IAILRKKAQLENLTVPNEVIVFIAD...
151                       EALKDILSANKAKVLNCTTIQEAVA...
152                       TEMSLPKIGEEFGGRDHTTVIHACE...
154                       /locus_tag="CLO1313_RS00015"
155                       /old_locus_tag="Clo1313_0...
157                       /locus_tag="CLO1313_RS00015"
158                       /old_locus_tag="Clo1313_0...
167                       /translation="MKIVCSKEQLM...
168                       GNDLELGIECLIDADILEKGSIVLN...
169                       HFELRGMPSDSFPSLPSIEKENMIK...
170                       AGNEITFVSIDGFRMALRKNFNNEG...
171                       NQILFEIGNCKVVSRLLEGEYLNYK...
172                       KYPVKFNIIDDKIIITSNTEIGAVR...
173                       DIYFNSSVGPCTIRPLEGDSFAYMI...
175                       /locus_tag="CLO1313_RS00020"
176                       /old_locus_tag="Clo1313_0...
178                       /locus_tag="CLO1313_RS00020"
179                       /old_locus_tag="Clo1313_0...
188                       /translation="MENIKINTEFI...
189                          ELRRGRKLRTGDIVEINDKKFQIV"
191                       /locus_tag="CLO1313_RS00025"
192                       /old_locus_tag="Clo1313_0...
194                       /locus_tag="CLO1313_RS00025"
195                       /old_locus_tag="Clo1313_0...
204                       /translation="MYIDRILLKNF...
205                       LCASGRSHRTSKDTELVNIDGTGFS...
206                       KKIGNLMGNLLAVIFSPEDILIIKE...
207                       RNMLLKEIQYKRNLLDTLEVWDYKM...
208                       EIMEIKYSPSVDLHDLSNPSEIKNE...
209                       LNLKMFGSQGQQRTSLLSLKLAEIE...
210                       RNVQTFITCTDKELFENRNFGDNLY...
212                       /locus_tag="CLO1313_RS00030"
213                       /old_locus_tag="Clo1313_0...
215                       /locus_tag="CLO1313_RS00030"
216                       /old_locus_tag="Clo1313_0...
225                       /translation="MFLHIGGDRVV...
226                       QTVSSDIPKSFIITETDKKSIIYLS...
227                                                KG"
230                       /locus_tag="CLO1313_RS00035"
231                       /old_locus_tag="Clo1313_0...
234                       /locus_tag="CLO1313_RS00035"

Remove duplicate values. It looks like each locus tag appears twice.

In [0]:
table_unique = table.drop_duplicates()

In [16]:
table_unique.head(20)

0
132                       /locus_tag="CLO1313_RS00010"
133                       /old_locus_tag="Clo1313_0...
145                       /translation="MNTQLNEIWQK...
146                       AVPAEFNKGILESRYQTLIKNAIKQ...
147                       PLSVLNPKYTFDTFVIGNSNRFAHA...
148                       GHYILEQNSSQRVLYVSSEKFTNEL...
149                       ERTEEEFFHTFNALYEANKQIILSS...
150                       IAILRKKAQLENLTVPNEVIVFIAD...
151                       EALKDILSANKAKVLNCTTIQEAVA...
152                       TEMSLPKIGEEFGGRDHTTVIHACE...
154                       /locus_tag="CLO1313_RS00015"
155                       /old_locus_tag="Clo1313_0...
167                       /translation="MKIVCSKEQLM...
168                       GNDLELGIECLIDADILEKGSIVLN...
169                       HFELRGMPSDSFPSLPSIEKENMIK...
170                       AGNEITFVSIDGFRMALRKNFNNEG...
171                       NQILFEIGNCKVVSRLLEGEYLNYK...
172                       KYPVKFNIIDDKIIITSNTEIGAVR...
173                       DIYFNSSVGPCTIRPLEGDSFAYMI...
175                       /locus_tag="CLO1313_RS00020"

In [0]:
table_unique = table_unique.reset_index(drop = True)

In [18]:
table_unique.head(20)

0
0                        /locus_tag="CLO1313_RS00010"
1                        /old_locus_tag="Clo1313_0...
2                        /translation="MNTQLNEIWQK...
3                        AVPAEFNKGILESRYQTLIKNAIKQ...
4                        PLSVLNPKYTFDTFVIGNSNRFAHA...
5                        GHYILEQNSSQRVLYVSSEKFTNEL...
6                        ERTEEEFFHTFNALYEANKQIILSS...
7                        IAILRKKAQLENLTVPNEVIVFIAD...
8                        EALKDILSANKAKVLNCTTIQEAVA...
9                        TEMSLPKIGEEFGGRDHTTVIHACE...
10                       /locus_tag="CLO1313_RS00015"
11                       /old_locus_tag="Clo1313_0...
12                       /translation="MKIVCSKEQLM...
13                       GNDLELGIECLIDADILEKGSIVLN...
14                       HFELRGMPSDSFPSLPSIEKENMIK...
15                       AGNEITFVSIDGFRMALRKNFNNEG...
16                       NQILFEIGNCKVVSRLLEGEYLNYK...
17                       KYPVKFNIIDDKIIITSNTEIGAVR...
18                       DIYFNSSVGPCTIRPLEGDSFAYMI...
19                       /locus_tag="CLO1313_RS00020"

The protein sequence always starts with "/translation" so I can use that to find the index of where protein sequence occurs. 

In [0]:
translation_index = table_unique[table_unique[0].str.contains('/translation')].index

In [20]:
translation_index

Int64Index([    2,    12,    21,    25,    34,    39,    53,    60,    68,
               74,
            ...
            25485, 25502, 25511, 25519, 25534, 25545, 25551, 25560, 25564,
            25569],
           dtype='int64', length=2917)

In [21]:
table_unique[0][12:19].values

array(['                     /translation="MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMT',
       '                     GNDLELGIECLIDADILEKGSIVLNSKMFGDIVRRLPDSEVLIEVKENNTVIIECDNS',
       '                     HFELRGMPSDSFPSLPSIEKENMIKVSQKAIRDMIRQTLFAVSMEGTRPILTGSLIEC',
       '                     AGNEITFVSIDGFRMALRKNFNNEGFSEFSVVVPAKTLSEIGKILQPVDEDIYIYSSQ',
       '                     NQILFEIGNCKVVSRLLEGEYLNYKSIIPPEYETSVRLRTEDLLSSLERASLITSDEK',
       '                     KYPVKFNIIDDKIIITSNTEIGAVREEIRVEVNGSNMEVGFNPRYFIEALRVIDDELV',
       '                     DIYFNSSVGPCTIRPLEGDSFAYMILPVRINK"'],
      dtype=object)

The locus_tag and old_locus_tag are at the end of the protein sequence so the index of end of protein sequence would be the index of the next protein sequence - 2.

In [0]:
new_df = pd.DataFrame(columns = ['old tag', 'new tag', 'sequence'])

In [0]:
sequences = []

In [24]:
table_unique[0][2:10].values

array(['                     /translation="MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINL',
       '                     AVPAEFNKGILESRYQTLIKNAIKQVTFKEYEIAFIVPSQENLNKLTKQTESAGNEDS',
       '                     PLSVLNPKYTFDTFVIGNSNRFAHAAALAVAEAPGKAYNPLFIYGGVGLGKTHLMHAI',
       '                     GHYILEQNSSQRVLYVSSEKFTNELINAIKDNRNEEFRSKYRNIDVLLIDDIQFIAGK',
       '                     ERTEEEFFHTFNALYEANKQIILSSDKPPKEISLEDRLRSRFEWGLIADMQAPDLETR',
       '                     IAILRKKAQLENLTVPNEVIVFIADKIASNIRELEGALNRVIAYSSLTENEITVELAS',
       '                     EALKDILSANKAKVLNCTTIQEAVARYFDIRPEEFKSKKRTRDIAFPRQIAMYLCREL',
       '                     TEMSLPKIGEEFGGRDHTTVIHACEKISEEIESNSETRRAVSEIKRNLLGK"'],
      dtype=object)

In [0]:
for i in range(len(translation_index)-1):
  sequences.append(list(table_unique[0][translation_index[i]:translation_index[i+1]-2].values))

In [26]:
len(sequences)

2916

Missing one protein bc the for loop range is -1.

In [27]:
translation_index[-1]

25569

In [28]:
table_unique[0][25569:25582]

25569                         /translation="MLRTYQPKKRQ...
25570              CONTIG      join(CP002416.1:1..3561619)
25571                                         ORIGIN      
25572            1 attttgtcaa ttattttgta tgcggaaata ttt...
25573           61 taaaatcaac gcaaaatcat aaacaataaa cca...
25574          121 caaggaaata attttacaaa ccaccgacaa acg...
25575          181 ccttattctt ttaatggttt tggaggaact tat...
25576          241 aaaaacttta ggactgctta aaaatgagct tac...
25577          301 gaccatcgat ccattgtcct tgacaggcaa tac...
25578          361 caacaaggga attcttgagt ccaggtatca aac...
25579          421 tacttttaag gaatacgaga ttgcatttat cgt...
25580          481 gacgaagcag accgagtccg ccggcaatga gga...
25581          541 gtacacgttt gacacttttg tcataggaaa cag...
Name: 0, dtype: object

In [0]:
sequences.append(list(table_unique[0][25569]))

In [30]:
len(sequences)

2917

In [32]:
sequences[:3]

[['                     /translation="MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINL',
  '                     AVPAEFNKGILESRYQTLIKNAIKQVTFKEYEIAFIVPSQENLNKLTKQTESAGNEDS',
  '                     PLSVLNPKYTFDTFVIGNSNRFAHAAALAVAEAPGKAYNPLFIYGGVGLGKTHLMHAI',
  '                     GHYILEQNSSQRVLYVSSEKFTNELINAIKDNRNEEFRSKYRNIDVLLIDDIQFIAGK',
  '                     ERTEEEFFHTFNALYEANKQIILSSDKPPKEISLEDRLRSRFEWGLIADMQAPDLETR',
  '                     IAILRKKAQLENLTVPNEVIVFIADKIASNIRELEGALNRVIAYSSLTENEITVELAS',
  '                     EALKDILSANKAKVLNCTTIQEAVARYFDIRPEEFKSKKRTRDIAFPRQIAMYLCREL',
  '                     TEMSLPKIGEEFGGRDHTTVIHACEKISEEIESNSETRRAVSEIKRNLLGK"'],
 ['                     /translation="MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMT',
  '                     GNDLELGIECLIDADILEKGSIVLNSKMFGDIVRRLPDSEVLIEVKENNTVIIECDNS',
  '                     HFELRGMPSDSFPSLPSIEKENMIKVSQKAIRDMIRQTLFAVSMEGTRPILTGSLIEC',
  '                     AGNEITFVSIDGFRMALRKNFNNEGFSEFSVVVPAKTLSEIGKILQ

The for loop worked. Now I just need to clean up the messy sequences.

In [0]:
import re

def clean_sequence(amino):
  a = ''.join(amino)
  b = re.sub(r"\s+", "", a)
  c = re.sub('/translation=','',b)
  d = re.sub('"', '', c)
  return d


In [0]:
clean_sequence_list = []
for s in sequences:
  clean_sequence_list.append(clean_sequence(s))

In [35]:
clean_sequence_list[:5]

['MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINLAVPAEFNKGILESRYQTLIKNAIKQVTFKEYEIAFIVPSQENLNKLTKQTESAGNEDSPLSVLNPKYTFDTFVIGNSNRFAHAAALAVAEAPGKAYNPLFIYGGVGLGKTHLMHAIGHYILEQNSSQRVLYVSSEKFTNELINAIKDNRNEEFRSKYRNIDVLLIDDIQFIAGKERTEEEFFHTFNALYEANKQIILSSDKPPKEISLEDRLRSRFEWGLIADMQAPDLETRIAILRKKAQLENLTVPNEVIVFIADKIASNIRELEGALNRVIAYSSLTENEITVELASEALKDILSANKAKVLNCTTIQEAVARYFDIRPEEFKSKKRTRDIAFPRQIAMYLCRELTEMSLPKIGEEFGGRDHTTVIHACEKISEEIESNSETRRAVSEIKRNLLGK',
 'MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMTGNDLELGIECLIDADILEKGSIVLNSKMFGDIVRRLPDSEVLIEVKENNTVIIECDNSHFELRGMPSDSFPSLPSIEKENMIKVSQKAIRDMIRQTLFAVSMEGTRPILTGSLIECAGNEITFVSIDGFRMALRKNFNNEGFSEFSVVVPAKTLSEIGKILQPVDEDIYIYSSQNQILFEIGNCKVVSRLLEGEYLNYKSIIPPEYETSVRLRTEDLLSSLERASLITSDEKKYPVKFNIIDDKIIITSNTEIGAVREEIRVEVNGSNMEVGFNPRYFIEALRVIDDELVDIYFNSSVGPCTIRPLEGDSFAYMILPVRINK',
 'MENIKINTEFIKLDQFLKWTKTVSMGSEAKLMIRSGLVKVNGEVELRRGRKLRTGDIVEINDKKFQIV',
 'MYIDRILLKNFRNYKDETIKFSKNLNIIYGQNAQGKTNIIEAVFLCASGRSHRTSKDTELVNIDGTGFSVLLDLESSEGRKKIEIDYECGKKKVVKINEIPLKKIG

Woohoo, so much better!

In [0]:
new_df = pd.DataFrame(columns = ['new tag','old tag','sequence'])

In [0]:
new_df['sequence'] = clean_sequence_list

In [38]:
new_df.head()

new tag old tag                                           sequence
0     NaN     NaN  MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINLAV...
1     NaN     NaN  MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMTGN...
2     NaN     NaN  MENIKINTEFIKLDQFLKWTKTVSMGSEAKLMIRSGLVKVNGEVEL...
3     NaN     NaN  MYIDRILLKNFRNYKDETIKFSKNLNIIYGQNAQGKTNIIEAVFLC...
4     NaN     NaN  MFLHIGGDRVVPVKNIIAILDMETTTISKDTKDFLAIAEEEGFIQT...

Next I will just need to add the tags/IDs.

In [39]:
type(translation_index)

pandas.core.indexes.numeric.Int64Index

In [0]:
old_tag_index = translation_index -1

In [41]:
old_tag_index[:5]

Int64Index([1, 11, 20, 24, 33], dtype='int64')

In [0]:
old_tag_list = []
for i in old_tag_index:
  old_tag_list.append(table_unique[0][i])

In [46]:
old_tag_list[:3]

['                     /old_locus_tag="Clo1313_0001"',
 '                     /old_locus_tag="Clo1313_0002"',
 '                     /old_locus_tag="Clo1313_0003"']

In [0]:
def old_tag(tag_name):
  a = re.sub(r"\s+", "", tag_name)
  b = re.sub('/old_locus_tag=','',a)
  c = re.sub('"', '', b)
  return c

In [48]:
old_tag(old_tag_list[0])

'Clo1313_0001'

In [49]:
clean_old_tag = []
for tag in old_tag_list:
  clean_old_tag.append(old_tag(tag))
  
print(clean_old_tag[:5])

['Clo1313_0001', 'Clo1313_0002', 'Clo1313_0003', 'Clo1313_0004', 'Clo1313_0005']


In [50]:
len(clean_old_tag)

2917

In [56]:
new_df['old tag'] = clean_old_tag
print(new_df.head())

  new tag       old tag                                           sequence
0     NaN  Clo1313_0001  MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINLAV...
1     NaN  Clo1313_0002  MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMTGN...
2     NaN  Clo1313_0003  MENIKINTEFIKLDQFLKWTKTVSMGSEAKLMIRSGLVKVNGEVEL...
3     NaN  Clo1313_0004  MYIDRILLKNFRNYKDETIKFSKNLNIIYGQNAQGKTNIIEAVFLC...
4     NaN  Clo1313_0005  MFLHIGGDRVVPVKNIIAILDMETTTISKDTKDFLAIAEEEGFIQT...


In [0]:
new_tag_index = translation_index -2

In [0]:
new_tag_list = []
for i in new_tag_index:
  new_tag_list.append(table_unique[0][i])

In [59]:
new_tag_list[:3]

['                     /locus_tag="CLO1313_RS00010"',
 '                     /locus_tag="CLO1313_RS00015"',
 '                     /locus_tag="CLO1313_RS00020"']

In [0]:
def new_tag(tag_name):
  a = re.sub(r"\s+", "", tag_name)
  b = re.sub('/locus_tag=','',a)
  c = re.sub('"', '', b)
  return c

In [63]:
new_tag(new_tag_list[0])

'CLO1313_RS00010'

In [64]:
clean_new_tag = []
for tag in new_tag_list:
  clean_new_tag.append(new_tag(tag))
  
print(clean_new_tag[:5])

['CLO1313_RS00010', 'CLO1313_RS00015', 'CLO1313_RS00020', 'CLO1313_RS00025', 'CLO1313_RS00030']


In [65]:
new_df['new tag'] = clean_new_tag
print(new_df.head())

           new tag  ...                                           sequence
0  CLO1313_RS00010  ...  MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINLAV...
1  CLO1313_RS00015  ...  MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMTGN...
2  CLO1313_RS00020  ...  MENIKINTEFIKLDQFLKWTKTVSMGSEAKLMIRSGLVKVNGEVEL...
3  CLO1313_RS00025  ...  MYIDRILLKNFRNYKDETIKFSKNLNIIYGQNAQGKTNIIEAVFLC...
4  CLO1313_RS00030  ...  MFLHIGGDRVVPVKNIIAILDMETTTISKDTKDFLAIAEEEGFIQT...

[5 rows x 3 columns]


In [0]:
pd.set_option('display.max_columns', None)

In [68]:
new_df.head()

new tag       old tag  \
0  CLO1313_RS00010  Clo1313_0001   
1  CLO1313_RS00015  Clo1313_0002   
2  CLO1313_RS00020  Clo1313_0003   
3  CLO1313_RS00025  Clo1313_0004   
4  CLO1313_RS00030  Clo1313_0005   

                                            sequence  
0  MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINLAV...  
1  MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMTGN...  
2  MENIKINTEFIKLDQFLKWTKTVSMGSEAKLMIRSGLVKVNGEVEL...  
3  MYIDRILLKNFRNYKDETIKFSKNLNIIYGQNAQGKTNIIEAVFLC...  
4  MFLHIGGDRVVPVKNIIAILDMETTTISKDTKDFLAIAEEEGFIQT...

What is the new tag and the amino acid sequence of Clo1313_0396?

In [69]:
new_df[new_df['old tag'] == 'Clo1313_0396']

new tag       old tag  \
386  CLO1313_RS02005  Clo1313_0396   

                                              sequence  
386  MTSEEIAKLCGVSRATVSRVINNSPNVKEETRQKILAVIKEKNYVP...

In [80]:
import numpy as np
new_df['sequence'].str.len().sum()/2917

341.7545423380185

The average protein sequence is 341.75 amino acids in this organism. 